Sources/MapboxMaps/Foundation/Events/Payload/NoPayload.swift

```diff
@@ -0,0 +1,5 @@
+import Foundation
+
+public struct NoPayload: Decodable {
```

Comment:
My recommendation is to use empty-case `enum` over `struct` as it generates less memory fingerprint. Little kind of phantom type.

Reply:
Instead of doing this, what do you think if we have a marker protocol called `MapEventType`, which has 2 conformers
- `HasPayloadMapEvent` represents an event with some payload.
- `EmptyPayloadMapEvent` represents an event with zero payload.

```
public protocol MapEventType {}

public struct HasPayloadMapEvent<Payload: Decodable>: MapEventType {
    private let event: MapboxCoreMaps.Event
    public var eventName: String { event.type }

    public lazy var payload: Payload! = try? event.typedPayload()

    internal init(event: MapboxCoreMaps.Event) {
        self.event = event
    }
}

public struct EmptyPayloadMapEvent:MapEventType {
    private let event: MapboxCoreMaps.Event
    public var eventName: String { event.type }

    internal init(event: MapboxCoreMaps.Event) {
        self.event = event
    }
}
```
then we would constraint `MapEvents.Event` to this `MapEventType`, and the subscription will have a benefit of knowing which type of event it is and whether it has payload or not.

Reply:
Sounds interesting!
But I do not quite understand how would the subscription know what type of payload it has?

Reply:
```
public extension MapEvents {
  /// Typed events emitted by the SDK
  struct Event<EventType: MapEventType> {
    /// The style has been fully loaded, and the map has rendered all visible tiles.
    public static var mapLoaded: Event<EmptyPayloadMapEvent> { .init(name: MapEvents.mapLoaded) }

    /// Describes an error that has occurred while loading the Map. The `type` property defines what resource could
    /// not be loaded and the `message` property will contain a descriptive error message.
    /// In case of `source` or `tile` loading errors, `source-id` will contain the id of the source failing.
    /// In case of `tile` loading errors, `tile-id` will contain the id of the tile.
    public static var mapLoadingError: Event<HasPayloadMapEvent<MapLoadingError>> { .init(name: MapEvents.mapLoadingError) }
}

@discardableResult
public func onNext<EventType>(event: MapEvents.Event<EventType>, handler: @escaping (EventType) -> Void) -> Cancelable {}

mapboxMap.onNext(event: .mapLoaded) { event in
  // event.eventName
  // event.payload <-- Value of type 'EmptyPayloadMapEvent' has no member 'payload'
}

mapboxMap.onNext(event: .mapLoadingError) { event in 
  // event.eventName
  // event.payload <-- MapLoadingError
}
```
It was something like this, do you think it makes more sense 🤔

Reply:
> It was something like this, do you think it makes more sense 🤔

Nice idea, I like it!
Should we discuss the naming?

Reply:
As we confirmed during an in-person discussion - changes to public protocols often result in a breaking change, so public protocols should only be introduced in situations where there's a reason for developers to implement the protocol. 
While this makes for a better API, doing this will require us to do expose a public protocol that's not supposed to be conformed to.